Improve gptune package

liuyangzhuan opened the pull request:
1. Address comments by tldahlgren in PR #26936
2. Add a variant +openmpi to allow installation using non openmpi modules

liuyangzhuan committed:
added package gptune with all its dependencies: adding py-autotune, pygmo, py-pyaml, py-autotune, py-gpy, py-lhsmdu, py-hpbandster, pagmo2, py-opentuner; modifying superlu-dist, py-scikit-optimize

liuyangzhuan committed:
adding gptune package

liuyangzhuan committed:
minor fix for macos spack test

liuyangzhuan committed:
update patch for py-scikit-optimize; update test files for gptune

liuyangzhuan committed:
merging conficts for superlu-dist

liuyangzhuan committed:
fixing gptune package style error

liuyangzhuan committed:
fixing unit tests

liuyangzhuan committed:
a few changes reviewed in the PR

liuyangzhuan committed:
merge conflict

liuyangzhuan committed:
improved gptune package.py with a few newly added/improved dependencies

liuyangzhuan committed:
fixed a few style errors

liuyangzhuan committed:
minor fix on package name py-pyro4

liuyangzhuan committed:
fixing more style errors

liuyangzhuan committed:
Update var/spack/repos/builtin/packages/py-scikit-optimize/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

liuyangzhuan committed:
resolved a few issues in the PR

liuyangzhuan committed:
Merge branch 'develop' of https://github.com/liuyangzhuan/spack into develop

liuyangzhuan committed:
fixing file permissions

liuyangzhuan committed:
a few minor changes

liuyangzhuan committed:
style correction

liuyangzhuan committed:
minor correction to jq package file

liuyangzhuan committed:
Update var/spack/repos/builtin/packages/py-pyro4/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

liuyangzhuan committed:
fixing a few issues in the PR

liuyangzhuan committed:
adding py-selectors34 required by py-pyro4

liuyangzhuan committed:
improved the superlu-dist package

liuyangzhuan committed:
improved the superlu-dist package

liuyangzhuan committed:
moree changes to gptune and py-selectors34 based on the PR

liuyangzhuan committed:
Update var/spack/repos/builtin/packages/py-selectors34/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

liuyangzhuan committed:
improved gptune package: 1. addressing comments of tldahlgren in PR 26936; 2. adding variant openmpi

liuyangzhuan committed:
fixing style issue of gptune

liuyangzhuan committed:
changing file mode

liuyangzhuan committed:
merging conflict

alalazo commented on var/spack/repos/builtin/packages/gptune/package.py:
```diff
@@ -17,8 +17,8 @@ class Gptune(CMakePackage):
     version('master', branch='master')
-
     variant('app', default=False, description='Build all HPC application examples')
+    variant('openmpi', default=True, description='MPI spawning-based interface')
```
Does this package depend explicitly on `openmpi` only?

liuyangzhuan replied:
I have changed the variant to "mpispawn" to clarify that openmpi is only needed if one would like to use the mpi spawning based feature in gptune.

alalazo commented on var/spack/repos/builtin/packages/gptune/package.py:
```diff
@@ -50,7 +50,10 @@ class Gptune(CMakePackage):
     depends_on('superlu-dist@develop', when='+app', type=('build', 'run'))
 
-    conflicts('openmpi@:3')
+    depends_on('openmpi@4:', when='+openmpi', type=('build', 'run'))
+    conflicts('mpich', when='+openmpi')
+    conflicts('spectrum-mpi', when='+openmpi')
+    conflicts('cray-mpich', when='+openmpi')
```
Why a `(build, run)` dependency for `openmpi`? Assuming `openmpi` is the only viable choice, I think this can be written in a more robust way as:
```suggestion
    depends_on('openmpi@4:', when='+openmpi')
    depends_on('mpi', when='+openmpi')
```
since the `mpi` dependency will rule out any other `mpi` provider.

liuyangzhuan replied:
@alalazo mpi is a mandatory requirement of gptune. 

If the variant +openmpi is enabled, gptune needs openmpi, otherwise, gptune can use mpich, spectrum-mpi, cray-mpich, openmpi or other mpi providers. 

So what's the appropriate way of specifying this?

liuyangzhuan replied:
But the variant should be named "openmpionly" to avoid confusion.

liuyangzhuan replied:
I've changed the variant name to "mpispawn"

alalazo reviewed (changes requested):
A comment and a couple of questions.

liuyangzhuan committed:
Merge branch 'develop' of https://github.com/spack/spack into develop

liuyangzhuan committed:
improved gptune package: add variant mpispawn which depends on openmpi; add variant superlu and hypre for installing the drivers; modified hypre package file to add a gptune variant

liuyangzhuan committed:
fixing style error

liuyangzhuan commented:
Hi, @alalazo I've made a few changes according to your comments:
1. the variant name is changed from "openmpi" to "mpispawn"
2. I modified hypre package to make it work with gptune. 
3. added variants "superlu" and "hypre"
4. simplified a few lines in the gptune package script

Can you take a look at it again?

liuyangzhuan committed:
corrected pddrive_spawn path in gptune test; enforcing gcc>7

liuyangzhuan committed:
fixing style error

liuyangzhuan committed:
setting environment variables when loading gptune

liuyangzhuan commented:
Hi, @adamjstewart, I've made some changes according to Issue #28116. Also there were some previous commits in this PR that address @alalazo comments but haven't been processed. Can you take a look altogether?

liuyangzhuan commented:
Hi, @adamjstewart, this PR has been open for quite a while. Has anyone in your team been able to look at it? Thanks!

adamjstewart commented:
@liuyangzhuan I don't have a team, I'm just a volunteer contributor. Maybe someone from @spack/merge-to-develop can review?

liuyangzhuan commented:
@adamjstewart Thanks for the clarification. Is there a way to join the spack contributors so that I can review PRs myself?

adamjstewart commented:
That's up to @tgamblin. Note that you can't review and merge your own PRs of course.

alalazo commented:
@spackbot maintainers

alalazo commented on var/spack/repos/builtin/packages/hypre/package.py:
```diff
@@ -247,6 +257,7 @@ def test(self):
             print('Skipping: HYPRE must be installed with +mpi to run tests')
             return
 
+        print(self._cached_tests_work_dir)
```
Is this a debug print?

liuyangzhuan replied:
That seems to be a debug print. I don't recall if I added it, or someone else. I think that line can be simply removed. 

@balay Could you take a look at the patch files for hypre I submitted in this PR? 

var/spack/repos/builtin/packages/hypre/ij_gptune.patch
var/spack/repos/builtin/packages/hypre/package.py

It's essentially added a few lines in ij.c to print out the runtime into the runlog, and get the input parameters from command line. Thanks!

liuyangzhuan replied:
@alalazo I have removed this debug printing line.

alalazo reviewed (dismissed):
I'm waiting for `hypre` maintainers review (apologies, but the bot didn't ping them). Can you point out in a comment over the patch directive:
1. The provenance of the patch
2. The reason why the patch is needed

?

liuyangzhuan committed:
removing debug print in hypre/package.py

liuyangzhuan committed:
adding superlu-dist v7.2.0; fixing an issue with CMAKE_INSTALL_LIBDIR

eugeneswalker commented:
We would like to include this in E4S 22.2. Hoping someone on Hypre team can greenlight the changes to hypre so we can merge this ASAP!

osborn9 reviewed (approved):
This looks good to me for the changes to hypre.

liuyangzhuan commented:
@osborn9  Thanks! @alalazo May we get the ball rolling now?

wspear reviewed (approved):
Looks good to me.
